cmd/auth-provider-gcp/plugin/plugin_test.go

```diff
@@ -0,0 +1,213 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package plugin
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	utilnet "k8s.io/apimachinery/pkg/util/net"
+	"k8s.io/cloud-provider-gcp/cmd/auth-provider-gcp/gcpcredential"
+	credentialproviderapi "k8s.io/kubelet/pkg/apis/credentialprovider"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+	"testing"
+)
+
+const (
+	serviceAccountsEndpoint = "/computeMetadata/v1/instance/service-accounts/"
+	defaultEndpoint         = "/computeMetadata/v1/instance/service-accounts/default/"
+	scopeEndpoint           = defaultEndpoint + "scopes"
+	emailEndpoint           = defaultEndpoint + "email"
+	tokenEndpoint           = defaultEndpoint + "token"
+	dummyToken              = "ya26.lots-of-indiscernible-garbage"
+	email                   = "1234@project.gserviceaccount.com"
+	expectedUsername        = "_token"
+)
+
+func hasURL(url string, response *credentialproviderapi.CredentialProviderResponse) bool {
+	if _, ok := response.Auth[url]; ok {
+		return true
+	}
+	return false
+}
+
+func usernameMatches(expectedUsername string, auth credentialproviderapi.AuthConfig) bool {
+	return auth.Username == expectedUsername
+}
+
+func passwordMatches(expectedPassword string, auth credentialproviderapi.AuthConfig) bool {
+	return auth.Password == expectedPassword
+}
+
+func TestContainerRegistry(t *testing.T) {
+	// Taken from from pkg/credentialprovider/gcp/metadata_test.go in k/k
+	registryURL := "container.cloud.google.com"
+	token := &gcpcredential.TokenBlob{AccessToken: dummyToken} // Fake value for testing.
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Only serve the URL key and the value endpoint
+		if scopeEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			w.Header().Set("Content-Type", "application/json")
+			fmt.Fprintf(w, `["%s.read_write"]`, gcpcredential.StorageScopePrefix)
+		} else if emailEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprint(w, email)
+		} else if tokenEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			w.Header().Set("Content-Type", "application/json")
+			bytes, err := json.Marshal(token)
+			if err != nil {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			fmt.Fprintln(w, string(bytes))
+		} else if serviceAccountsEndpoint == r.URL.Path {
+			w.WriteHeader(http.StatusOK)
+			fmt.Fprintln(w, "default/\ncustom")
+		} else {
+			http.Error(w, "", http.StatusNotFound)
+		}
+	}))
+	defer server.Close()
+	// Make a transport that reroutes all traffic to the example server
+	transport := utilnet.SetTransportDefaults(&http.Transport{
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return url.Parse(server.URL + req.URL.Path)
+		},
+	})
+	provider := MakeRegistryProvider(transport)
+	response, err := GetResponse(provider)
+	if err != nil {
+		t.Errorf("Unexpected error while getting response: %s", err.Error())
```

Comment:
Use `t.Fatalf` for unrecoverable all errors, like this one.

Reply:
Gotcha, done